runtime/calamari/src/fee.rs

```diff
@@ -67,3 +47,119 @@ impl WeightToFeePolynomial for WeightToFee {
+			assertions()
+		});
+	}
+
+	// update based on runtime impl.
+	fn runtime_multiplier_update(fm: Multiplier) -> Multiplier {
+		TargetedFeeAdjustment::<
+			Runtime,
+			TargetBlockFullness,
+			AdjustmentVariable,
+			MinimumMultiplier,
+		>::convert(fm)
+	}
+
+	fn fetch_kma_price() -> Result<f32, &'static str> {
+		let body = reqwest::blocking::get(
+			"https://api.coingecko.com/api/v3/simple/price?ids=calamari-network&vs_currencies=usd",
+		)
+		.unwrap();
+		let json_reply: serde_json::Value = serde_json::from_reader(body).unwrap();
+		if let Some(price) = json_reply["calamari-network"]["usd"].as_f64() {
+			// CG API return: {"calamari-network":{"usd": 0.01092173}}
+			Ok(price as f32)
+		} else {
+			Err("KMA price not found in reply from Coingecko. API changed? Check https://www.coingecko.com/en/api/documentation")
+		}
+	}
+
+	// Consider the daily cost to fully congest our network to be defined as:
+	// `target_daily_congestion_cost_usd = inclusion_fee * blocks_per_day * kma_price`
+	// Where:
+	// `inclusion_fee = fee_adjustment * (weight_to_fee_coeff * (block_weight ^ degree)) + base_fee + length_fee`
+	// Where:
+	// `fee_adjustment` and `weight_to_fee_coeff` are configurable in a runtime via `FeeMultiplierUpdate` and `WeightToFee`
+	// `fee_adjustment` is also variable depending on previous block's fullness
+	// We are also assuming `length_fee` is negligible for small TXs like a remark or a transfer.
+	// This test loops 1 day of parachain blocks (7200) and calculates accumulated fee if every block is almost full
+	#[test]
+	fn congested_chain_simulation() {
+		// Configure the target cost depending on the current state of the network.
+		let target_daily_congestion_cost_usd = 250000;
+		let kma_price = fetch_kma_price().unwrap();
+		println!("KMA/USD price as read from CoinGecko = {}", kma_price);
+		let target_daily_congestion_cost_kma =
+			(target_daily_congestion_cost_usd as f32 * kma_price * KMA as f32) as u128;
+
+		// `cargo test --package calamari-runtime --lib -- fee::multiplier_tests::congested_chain_simulation --exact --nocapture` to get some insight.
+		// almost full. The entire quota of normal transactions is taken.
+		let block_weight = BlockWeights::get()
+			.get(DispatchClass::Normal)
+			.max_total
+			.unwrap() - 10;
+
+		let base_fee = <Runtime as pallet_transaction_payment::Config>::WeightToFee::calc(
+			&frame_support::weights::constants::ExtrinsicBaseWeight::get(),
+		);
+
+		run_with_system_weight(block_weight, || {
+			// initial value configured on module
+			let mut fee_adjustment = Multiplier::one();
+			assert_eq!(fee_adjustment, TransactionPayment::next_fee_multiplier());
+			let mut accumulated_fee: u128 = 0;
+			// Simulates 1 day of parachain blocks (12 seconds each)
+			for iteration in 0..7200 {
+				let next = runtime_multiplier_update(fee_adjustment);
+				// if no change, panic. This should never happen in this case.
+				if fee_adjustment == next {
```

Comment:
can next not be returned as something smaller than the last iteration?

Reply:
No, but let's just put a sanity check anyway.